src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -199,23 +212,174 @@ section torsion_polynomial
+by { rw [nonsingular_iff', sub_ne_zero, ← @sub_ne_zero _ _ y], congr' 3; ring1 }
+
+@[simp] lemma nonsingular_zero : W.nonsingular 0 0 ↔ W.a₃ ≠ 0 ∨ W.a₄ ≠ 0 :=
+by rw [nonsingular, C_0, eval_polynomial_X_zero, neg_ne_zero, eval_polynomial_Y_zero, or_comm]
+
+lemma nonsingular_iff_variable_change (x y : R) :
+  W.nonsingular x y ↔ (W.variable_change 1 x 0 y).nonsingular 0 0 :=
+begin
+  rw [nonsingular_iff', ← neg_ne_zero, or_comm, nonsingular_zero, variable_change_a₃,
+      variable_change_a₄, inv_one, units.coe_one],
+  congr' 3,
+  all_goals { ring1 }
+end
+
+lemma nonsingular_zero_of_Δ_ne_zero (h : W.equation 0 0) (hΔ : W.Δ ≠ 0) : W.nonsingular 0 0 :=
+by { simp only [equation_zero, nonsingular_zero] at *, contrapose! hΔ, simp [h, hΔ] }
+
+/-- A Weierstrass curve is nonsingular at every point if its discriminant is non-zero. -/
+lemma nonsingular_of_Δ_ne_zero {x y : R} (h : W.equation x y) (hΔ : W.Δ ≠ 0) : W.nonsingular x y :=
+(W.nonsingular_iff_variable_change x y).mpr $
+  nonsingular_zero_of_Δ_ne_zero _ ((W.equation_iff_variable_change x y).mp h) $
+by rwa [variable_change_Δ, inv_one, units.coe_one, one_pow, one_mul]
+
+lemma polynomial_eq : W.polynomial = cubic.to_poly
+  ⟨0, 1, cubic.to_poly ⟨0, 0, W.a₁, W.a₃⟩, cubic.to_poly ⟨-1, -W.a₂, -W.a₄, -W.a₆⟩⟩ :=
+by { simp only [polynomial, cubic.to_poly, C_0, C_1, C_neg, C_add, C_mul], ring1 }
+
+lemma polynomial_ne_zero [nontrivial R] : W.polynomial ≠ 0 :=
+by { rw [polynomial_eq], exact cubic.ne_zero_of_b_ne_zero one_ne_zero }
+
+lemma polynomial_degree [nontrivial R] : W.polynomial.degree = 2 :=
+by { rw [polynomial_eq], exact cubic.degree_of_b_ne_zero' one_ne_zero }
+
+lemma polynomial_nat_degree [nontrivial R] : W.polynomial.nat_degree = 2 :=
+by { rw [polynomial_eq], exact cubic.nat_degree_of_b_ne_zero' one_ne_zero }
+
+lemma polynomial_monic : W.polynomial.monic :=
+by { nontriviality R, simpa only [polynomial_eq] using cubic.monic_of_b_eq_one' }
+
+lemma polynomial_irreducible [nontrivial R] [no_zero_divisors R] : irreducible W.polynomial :=
+begin
+  by_contra h,
+  rcases (W.polynomial_monic.not_irreducible_iff_exists_add_mul_eq_coeff W.polynomial_nat_degree).mp
+          h with ⟨f, g, h0, h1⟩,
+  simp only [polynomial_eq, cubic.coeff_eq_c, cubic.coeff_eq_d] at h0 h1,
+  apply_fun degree at h0 h1,
+  rw [cubic.degree_of_a_ne_zero' $ neg_ne_zero.mpr $ one_ne_zero' R, degree_mul] at h0,
+  replace h1 := h1.symm.le.trans cubic.degree_of_b_eq_zero',
+  contrapose! h1,
+  rcases nat.with_bot.add_eq_three_iff.mp h0.symm with ⟨hf, hg⟩ | ⟨hf, hg⟩ | ⟨hf, hg⟩ | ⟨hf, hg⟩,
+  any_goals { rw [degree_add_eq_left_of_degree_lt]; simp only [hf, hg]; dec_trivial },
+  any_goals { rw [degree_add_eq_right_of_degree_lt]; simp only [hf, hg]; dec_trivial }
+end
+
+/-- The coordinate ring $R[W] := R[X, Y] / \langle W(X, Y) \rangle$ of `W`. -/
+@[reducible] def coordinate_ring : Type u := adjoin_root W.polynomial
+
+instance {F : Type u} [field F] (W : weierstrass_curve F) : is_domain W.coordinate_ring :=
```

Comment:
```suggestion
instance [is_domain R] [normalized_gcd_monoid R] : is_domain W.coordinate_ring :=
```
automatically applies to fields.

Reply:
A `field` is a `normalized_gcd_monoid` only if we have `decidable_eq` - will this not be a problem?

Reply:
Indeed, [comm_group_with_zero.normalized_gcd_monoid](https://leanprover-community.github.io/mathlib_docs/algebra/gcd_monoid/basic.html#comm_group_with_zero.normalized_gcd_monoid) requires decidable_eq. Even though
```
instance coordinate_ring.is_domain_of_field {F : Type u} [field F] (W : weierstrass_curve F) :
  is_domain W.coordinate_ring := infer_instance
```
it's likely due to thsi [leakage problem](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/.60classical.60.20attribute.20leakage), and it will fail outside of this section. So probably we should do this instead:
```
instance [is_domain R] [normalized_gcd_monoid R] : is_domain W.coordinate_ring :=
(ideal.quotient.is_domain_iff_prime _).mpr $
by simpa only [ideal.span_singleton_prime W.polynomial_ne_zero, ← gcd_monoid.irreducible_iff_prime]
    using W.polynomial_irreducible

instance coordinate_ring.is_domain_of_field {F : Type u} [field F] (W : weierstrass_curve F) :
  is_domain W.coordinate_ring := by { classical, apply_instance }
```